### PR TITLE
APPS-2904: metrics service

### PR DESCRIPTION
**Pull request by rostikjoystick**
**YouTrack task**: [#APPS-2925](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-2925)

**Description List**:
* Collect events locally
* Events batch network call

### CODE REVIEW

File: app/src/main/java/org/stepic/droid/analytic/AnalyticImpl.kt [hunk 1]
```diff
@@ -123,6 +126,11 @@ constructor(
             val bundle = bundleOf(*analyticEvent.params.map { (a, b) -> a to b }.toTypedArray())
             firebaseAnalytics.logEvent(analyticEvent.name, bundle)
         }
+
+        if (AnalyticSource.STEPIK_API in analyticEvent.sources) {
+            val properties = analyticEvent.params
+            stepikAnalytic.logEvent(analyticEvent.name, properties)
```
Comment (eadm) [inline app/src/main/java/org/stepic/droid/analytic/AnalyticImpl.kt:132]: is `analyticEvent.params` equals to event type that we should send to API?

File: app/src/main/java/org/stepic/droid/base/App.kt [hunk 2]
```diff
@@ -92,6 +98,34 @@ class App : MultiDexApplication() {
         init()
     }
 
+    private fun fetchProcessName(): String {
```
Comment (eadm) [inline app/src/main/java/org/stepic/droid/base/App.kt:101]: why this code is not equals to code in `Context:: isMainProcess` extension?

File: app/src/main/java/org/stepic/droid/ui/activities/MainFeedActivity.kt [hunk 3]
```diff
@@ -152,6 +165,11 @@ class MainFeedActivity : BackToExitActivityWithSmartLockBase(),
 
         initGoogleApiClient(true)
 
+        // TODO Testing analytics
```
Comment (eadm) [inline app/src/main/java/org/stepic/droid/ui/activities/MainFeedActivity.kt:168]: commented code

File: app/src/main/java/org/stepik/android/cache/analytic/AnalyticCacheDataSourceImpl.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import org.stepic.droid.storage.dao.IDao
+import org.stepik.android.data.analytic.source.AnalyticCacheDataSource
+import org.stepik.android.model.analytic.AnalyticLocalEvent
+import timber.log.Timber
+import javax.inject.Inject
+
+class AnalyticCacheDataSourceImpl
+@Inject
+constructor(
+    private val analyticDao: IDao<AnalyticLocalEvent>
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/cache/analytic/AnalyticCacheDataSourceImpl.kt:14]: won't we have issues by working with the same DB from different processes?

File: app/src/main/java/org/stepik/android/data/analytic/repository/AnalyticRepositoryImpl.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        analyticCacheDataSource.logEvent(analyticEvent)
+
+    // TODO Clean up flush
+    override fun flushEvents(): Completable =
+        analyticCacheDataSource
+            .getAllEvents()
+            .flatMapCompletable { events ->
+                val batchEvents = analyticBatchMapper.mapLocalToBatchEvents(events)
+                Completable.complete()
+//                analyticRemoteDataSource.flushEvents(batchEvents)
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/data/analytic/repository/AnalyticRepositoryImpl.kt:28]: not implemented

File: app/src/main/java/org/stepik/android/domain/analytic/interactor/AnalyticInteractor.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    companion object {
+        private const val TIMESTAMP_KEY = "timestamp"
+        private const val BUNDLE_MAP_KEY = "mMap"
+    }
+    fun logEvent(eventName: String, properties: Bundle): Completable {
+        val timeStamp = properties.getLong(TIMESTAMP_KEY, 0L)
+        properties.remove(TIMESTAMP_KEY)
+        val analyticEvent = AnalyticLocalEvent(
+            name = eventName,
+            eventData = Gson().toJsonTree(properties).asJsonObject[BUNDLE_MAP_KEY],
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/analytic/interactor/AnalyticInteractor.kt:24]: 🤔

File: app/src/main/java/org/stepik/android/view/analytic/AnalyticContentProvider.kt [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        const val LOG = "log"
+    }
+
+    private lateinit var component: AnalyticComponent
+
+    private val compositeDisposable = CompositeDisposable()
+
+    @Inject
+    @AnalyticFlush
+    lateinit var analyticsObservable: Observable<Unit>
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/view/analytic/AnalyticContentProvider.kt:38]: Why these observables are exposed?

File: app/src/main/java/org/stepik/android/view/analytic/AnalyticContentProvider.kt [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            )
+    }
+
+    private fun subscribeForFlushUpdates() {
+        compositeDisposable += analyticsObservable
+            .subscribeOn(backgroundScheduler)
+            .subscribeBy(
+                onNext = {
+                    analyticInteractor
+                        .flushEvents()
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/view/analytic/AnalyticContentProvider.kt:93]: flatMap?

File: app/src/main/java/org/stepik/android/view/analytic/AnalyticContentProvider.kt [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                        .subscribeOn(backgroundScheduler)
+                        .subscribe()
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/view/analytic/AnalyticContentProvider.kt:95]: Errors not handled

File: app/src/stage/java/org/stepic/droid/util/DebugToolsHelper.kt [hunk 10]
```diff
@@ -37,6 +37,24 @@ object DebugToolsHelper {
         }
     }
 
+    /***
+     *  This function can be used to debug non-main processes
+     */
+    fun initDebugTools(context: Context) {
```
Comment (eadm) [inline app/src/stage/java/org/stepic/droid/util/DebugToolsHelper.kt:43]: it seems that we can add check for application context and reduce duplication

File: model/src/main/java/org/stepik/android/model/analytic/AnalyticLocalEvent.kt [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package org.stepik.android.model.analytic
+
+import android.os.Parcel
+import android.os.Parcelable
+import com.google.gson.JsonElement
+
+data class AnalyticLocalEvent(
```
Comment (eadm) [inline model/src/main/java/org/stepik/android/model/analytic/AnalyticLocalEvent.kt:7]: parcelize?

File: app/src/main/java/org/stepic/droid/ui/activities/MainFeedActivity.kt (hunk 3 above)
Comment (rostikjoystick) [reply app/src/main/java/org/stepic/droid/ui/activities/MainFeedActivity.kt:159]: For testing purposes, will remove after finishing the feature.

File: app/src/main/java/org/stepic/droid/storage/operations/DatabaseFacade.kt [hunk 12]
```diff
@@ -27,6 +28,7 @@ import org.stepik.android.domain.last_step.model.LastStep
 import org.stepik.android.domain.user_courses.model.UserCourse
 import org.stepik.android.model.*
 import org.stepik.android.model.Unit
+import org.stepik.android.cache.analytic.model.AnalyticLocalEvent
```
Comment (eadm) [inline app/src/main/java/org/stepic/droid/storage/operations/DatabaseFacade.kt:31]: not used

File: app/src/main/java/org/stepic/droid/ui/activities/MainFeedActivity.kt (hunk 3 above)
Comment (eadm) [reply app/src/main/java/org/stepic/droid/ui/activities/MainFeedActivity.kt:160]: still needed?

File: app/src/main/java/org/stepik/android/cache/analytic/converter/JsonElementConverter.kt [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package org.stepik.android.cache.analytic.converter
+
+import androidx.room.TypeConverter
+import com.google.gson.Gson
+import com.google.gson.JsonElement
+
+class JsonElementConverter {
+    @TypeConverter
+    fun stringToJsonElement(value: String?): JsonElement =
+        Gson().toJsonTree(value)
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/cache/analytic/converter/JsonElementConverter.kt:10]: don't we want to use one from DI?

File: app/src/main/java/org/stepik/android/domain/analytic/TestAnalyticEvent.kt [hunk 14]
```diff
@@ -0,0 +1,34 @@
+package org.stepik.android.domain.analytic
+
+import org.stepic.droid.util.DateTimeHelper
+import org.stepik.android.domain.base.analytic.AnalyticEvent
+import org.stepik.android.domain.base.analytic.AnalyticSource
+import org.stepik.android.domain.course.analytic.CourseViewSource
+import java.util.EnumSet
+
+class TestAnalyticEvent(
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/analytic/TestAnalyticEvent.kt:9]: to remove

File: app/src/main/java/org/stepik/android/remote/analytic/AnalyticRemoteDataSourceImpl.kt [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+import org.stepik.android.remote.analytic.model.AnalyticBatchRequest
+import org.stepik.android.remote.analytic.service.AnalyticService
+import javax.inject.Inject
+
+class AnalyticRemoteDataSourceImpl
+@Inject
+constructor(
+    private val analyticService: AnalyticService
+) : AnalyticRemoteDataSource {
+    override fun flushEvents(events: List<AnalyticBatchEvent>): Completable =
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/remote/analytic/AnalyticRemoteDataSourceImpl.kt:15]: need to check for empty list

File: app/src/main/java/org/stepik/android/data/analytic/repository/AnalyticRepositoryImpl.kt [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    private val analyticBatchMapper: AnalyticBatchMapper,
+    private val analyticRemoteDataSource: AnalyticRemoteDataSource,
+    private val analyticCacheDataSource: AnalyticCacheDataSource
+) : AnalyticRepository {
+    override fun logEvent(analyticEvent: AnalyticLocalEvent): Completable =
+        analyticCacheDataSource.logEvent(analyticEvent)
+
+    override fun flushEvents(): Completable =
+        analyticCacheDataSource
+            .getAllEvents()
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/data/analytic/repository/AnalyticRepositoryImpl.kt:23]: let chunk with 100 elements

File: app/src/main/java/org/stepik/android/domain/analytic/interactor/AnalyticInteractor.kt [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+class AnalyticInteractor
+@Inject
+constructor(
+    private val analyticRepository: AnalyticRepository,
+    private val gson: Gson
+) {
+    companion object {
+        private const val TIMESTAMP_KEY = "timestamp"
+    }
+    fun logEvent(eventName: String, bundle: Bundle): Completable {
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/analytic/interactor/AnalyticInteractor.kt:21]: mapping should be in background

File: app/src/main/java/org/stepik/android/domain/course/analytic/CoursePreviewScreenOpenedAnalyticEvent.kt [hunk 18]
```diff
@@ -24,4 +26,7 @@ class CoursePreviewScreenOpenedAnalyticEvent(
             PARAM_IS_PAID to course.isPaid,
             PARAM_SOURCE to source.name
         ) + source.params.mapKeys { "${PARAM_SOURCE}_$it" }
+
+    override val sources: EnumSet<AnalyticSource> =
+        EnumSet.complementOf(EnumSet.of(AnalyticSource.STEPIK_API))
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course/analytic/CoursePreviewScreenOpenedAnalyticEvent.kt:31]: that should be by default in base class

File: app/src/main/java/org/stepik/android/domain/course/analytic/batch/CourseCardSeenAnalyticBatchEvent.kt [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        private const val PARAM_COURSE = "course"
+        private const val PARAM_SOURCE = "source"
+    }
+
+    override val name: String =
+        "Course card seen"
+
+    override val params: Map<String, Any> =
+        mapOf(
+            TIMESTAMP to DateTimeHelper.nowUtc(),
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course/analytic/batch/CourseCardSeenAnalyticBatchEvent.kt:26]: timestamp should be set up in analytics provider

File: app/src/main/java/org/stepik/android/domain/course/analytic/batch/CoursePreviewScreenOpenedAnalyticBatchEvent.kt [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        "Course preview screen opened"
+
+    override val params: Map<String, Any> =
+        mapOf(
+            TIMESTAMP to DateTimeHelper.nowUtc(),
+            PARAM_SOURCE to source.name,
+            DATA to mapOf(
+                PARAM_COURSE to course.id,
+                PARAM_TITLE to course.title.toString(),
+                PARAM_IS_PAID to course.isPaid
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course/analytic/batch/CoursePreviewScreenOpenedAnalyticBatchEvent.kt:34]: we don't have such fields in analytic 🤔

File: app/src/main/java/org/stepik/android/remote/analytic/service/AnalyticService.kt [hunk 21]
```diff
@@ -0,0 +1,13 @@
+package org.stepik.android.remote.analytic.service
+
+import io.reactivex.Completable
+import org.stepik.android.remote.analytic.model.AnalyticBatchRequest
+import retrofit2.http.Body
+import retrofit2.http.POST
+
+interface AnalyticService {
+    // TODO Commented for testing purposes, in order not to spam our backend
+//    @POST("api/metrics/batch")
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/remote/analytic/service/AnalyticService.kt:10]: not to forget

**Pull request merged**